feat(highlight): tweak default color scheme

echasnovski opened the pull request:
This is a follow up to #26334.

Related to #26369. Related to #26378.
Resolves #26385. Resolves #26389. Resolves #26468.

It mostly follows a list from the first task of #26369.

------

Here are the important points which might deserve extra attention during review:

1. Lightness values of `Nvim{Dark,Light}Grey1` (darkest/lightest) and `Nvim{Dark,Light}Grey2` (`Normal` highlight group) are made more extreme to increase contrast ratio. This makes background darker/lighter with foreground lighter/darker. This also prompted non-grey colors to shift a bit as they have lightness taken from `Normal` group.

    I **really** think that current values have the most possible compromise between "much contrast" and "not too much contrast" camps. All contrast ratio values passed previously (except objectively hard "saturated dark cyan") and now they pass even better (for example, `Comment` is now closer to passing with AAA than with AA).

1. `cterm` values are changed to use colors 0-6 and 9-15 to more utilize terminal colors and be treated more like "emergency exit" values (especially after #26407 is merged). Both dark/light variant use only 8 colors: one background, one foreground, and 6 colors (red, green, yellow, blue, magenta, cyan). Not using shades of grey (7 for dark-ish foreground and 8 for grey) allows a UI which is more robust towards terminal color palette not designed for these greys to be used together with non-grey colors.

    Currently the only assumption about terminal palette is that "Background color 0 can be used as background; foreground and colors can be used as foreground". This essentially limits the possible fg-bg color combinations to bg-{fg,red,green,yellow,blue,magenta,cyan} and fg-bg (inverted `Normal`).

    Most UI decisions are similar to ones from 'quiet' color scheme in case of only 16 colors available. Colors are translated from the corresponding `gui*` values: 9-14 for dark variant and 1-6 for light (to possibly have more contrast).

1. All highlight groups are reworked to follow "add attribute only if it is intended to be different". This includes two cases:
    - Add `guifg`/`ctermfg` if it is meant to have different foreground. For example, most "basic" syntax groups (`Constant`, `Operator`, `PreProc`, `Type`, `Delimiter`) are left without definition. This means that they will inherit the foreground, which is usually OK but can lead to not really good result inside markdown code blocks (falling back to `@text.literal` which links to `Comment`). @clason, this is probably better be addressed in tree-sitter highlight groups rework, correct?

    - Add `guibg`/`ctermbg` if it is meant to have different background. This resolves some issues with `DiagnosticFloating*`.

1. Specific highlight group decisions which might need discussion:

    1. `CurSearch` is changed to actually be different from `Search` (with antipodal lightness value of yellow as background). This adds another breaking behavior for color scheme authors, but seems to be worth having in default color scheme.

    1. `WinSeparator` was previously linked to a `VertSplit` for compatibility with Vim's color scheme. As `VertSplit` is deprecated since 0.8, I decided to revert the link. Even now I think it is a reasonable thing to do, but this might need extra discussion. Here are possible actions:
        - Leave the current `VertSplit -> WinSeparator -> Normal` linking dependency.
        - Revert to have `WinSeparator -> VertSplit -> Normal`.
        - Remove `VertSplit` entirely from default color scheme. Having it linked to `WinSeparator` is beneficial only to allow Neovim targeted color schemes work in Vim, which is rarely the case. So removing it completely might be reasonable here.

    1. As current default `EndOfBuffer` is already quite non intrusive, I don't think having it being separate from `Whitespace` adds benefit to address [this comment](https://github.com/neovim/neovim/issues/21694#issuecomment-1575756133). So nothing has changed here in this follow up.

    1. `PmenuSel` (current selection in pop-up menu) in the #26334 was changed to include `blend=0`. This is because if pop-up menu is transparent, current selection also amplifies the background text which makes it very uncomfortable to read. As this is a breaking change, it might be worth a discussion of whether this is necessary.

    1. `FloatShadow` and `FloatShadowThrough` are changed to have `Grey4` color implementing "soft shadow". This is mostly to make them visible with both dark and light color scheme variants.

    1. `FloatFooter` is changed to link to `FloatTitle` (instead of previous `Title`).

------

Information about new color scheme:

<details><summary>Palette</summary>

- `#004c73` (Dark blue)
- `#007373` (Dark cyan)
- `#005523` (Dark green)
- `#07080d` (Dark grey1)
- `#14161b` (Dark grey2)
- `#2c2e33` (Dark grey3)
- `#4f5258` (Dark grey4)
- `#470045` (Dark magenta)
- `#590008` (Dark red)
- `#6b5300` (Dark yellow)
- `#a6dbff` (Light blue)
- `#8cf8f7` (Light cyan)
- `#b3f6c0` (Light green)
- `#eef1f8` (Light grey1)
- `#e0e2ea` (Light grey2)
- `#c4c6cd` (Light grey3)
- `#9b9ea4` (Light grey4)
- `#ffcaff` (Light magenta)
- `#ffc0b9` (Light red)
- `#fce094` (Light yellow)

</details>

<details><summary>Contrast ratio values</summary>

`dark_*` are for dark color scheme, `ligth_*` - for light.

```
dark_blue         = 12.2
dark_comment      = 6.7
dark_comment_cur  = 5.1
dark_comment_vis  = 2.9
dark_cur_line     = 10.5
dark_cyan         = 14.6
dark_green        = 14.5
dark_magenta      = 13.1
dark_normal       = 14
dark_red          = 11.6
dark_visual       = 6.1
dark_yellow       = 14
light_blue        = 7.1
light_comment     = 6.1
light_comment_cur = 4.6
light_comment_vis = 2.9
light_cur_line    = 10.6
light_cyan        = 4.4
light_green       = 7
light_magenta     = 12
light_normal      = 14
light_red         = 11.4
light_visual      = 6.7
light_yellow      = 5.7
```

</details>

<details><summary>Screenshots</summary>

With `termguicolors`:

![default-cs2_dark_tgc](https://github.com/neovim/neovim/assets/24854248/5a730823-7579-49cf-bc1d-9c9e16cb488a)
![default-cs2_dark_tgc_diff](https://github.com/neovim/neovim/assets/24854248/5b56ca08-e193-4e26-97a7-1fe8e38586eb)
![default-cs2_light_tgc](https://github.com/neovim/neovim/assets/24854248/7a1939f0-2196-43fe-8c7c-04fd45056611)
![default-cs2_light_tgc_diff](https://github.com/neovim/neovim/assets/24854248/17337301-5116-4209-bb83-eaea87b63e18)

With `notermguicolors` and [VGA color palette](https://en.wikipedia.org/wiki/ANSI_escape_code#3-bit_and_4-bit) (for popular terminal palettes it should look better):

![default-cs2_dark_notgc](https://github.com/neovim/neovim/assets/24854248/86cccdd6-a632-487a-8b8a-27da64921fed)
![default-cs2_dark_notgc_diff](https://github.com/neovim/neovim/assets/24854248/1cdb3785-3dae-4778-be4b-054be259b7da)
![default-cs2_light_notgc](https://github.com/neovim/neovim/assets/24854248/695c6c34-689e-452b-b485-2824b9a99540)
![default-cs2_light_notgc_diff](https://github.com/neovim/neovim/assets/24854248/91a3fa98-6ef3-4fcb-80ec-4c1e243c87fd)

</details>

echasnovski committed:
feat(highlight): tweak default color scheme

Problem: Updating default color scheme produced some feedback.

Solution: Address the feedback.

Outline of the changes:

- Colors `Grey1` and `Grey2` are made a little bit more extreme (dark -
  darker, light - lighter) to increase overall contrast.

- `gui` colors are treated as base with `cterm` colors falling back to
  using 0-15 colors which come from terminal emulator.

- Update highlight group definition to not include attribute definition
  if it is intended to staty uncolored.

- Tweak some specific highlight groups.

- Add a list of Neovim specific highlight groups which are now defined
  differently in a breaking way.

- Minor tweaks in several other places related to default color scheme.

echasnovski commented:
I think this is ready for review. Will fix failing screenshot tests after highlight group values are approved.

gpanders commented:
>Currently the only assumption about terminal palette is that "Background color 0 can be used as background; foreground and colors can be used as foreground".

I don't think this is a valid assumption. Color 0 is not necessarily (nor even _usually_) the same as the background color.

echasnovski commented:
> > Currently the only assumption about terminal palette is that "Background color 0 can be used as background; foreground and colors can be used as foreground".
> 
> I don't think this is a valid assumption. Color 0 is not necessarily (nor even _usually_) the same as the background color.

Right. I mostly meant "background color can be used ...". Here colors 0 and 15 are suggested to be used if using `NONE` / `fg` / `bg` as colors is not possible.

Having set `ctermfg=0 ctermbg=15` for light `Normal` and `ctermfg=15 and ctermbg=0` for dark `Normal` is mostly because otherwise there is an error after `:colorscheme vim`: `E420: BG color unknown` (due to at least [these lines](https://github.com/neovim/neovim/blob/1907abb4c27857fe7f4e7394f32e130f9955a2e7/runtime/colors/vim.vim#L17-L18)).

gpanders commented:
>Having set ctermfg=0 ctermbg=15 for light Normal and ctermfg=15 and ctermbg=0 for dark Normal is mostly because otherwise there is an error after :colorscheme vim: E420: BG color unknown (due to at least [these lines](https://github.com/neovim/neovim/blob/1907abb4c27857fe7f4e7394f32e130f9955a2e7/runtime/colors/vim.vim#L17-L18)).

Just opened #26541 to see if we can get around this limitation.

gpanders commented:
#26541 is not the right solution. But the only highlight groups in the `vim` colorscheme that use `fg` and `bg` for the cterm colors are `Cursor` and `lCursor`. But we can just use `NONE` for those:

```diff
diff --git a/runtime/colors/vim.vim b/runtime/colors/vim.vim
index 3418abe42..0f6c5021d 100644
--- a/runtime/colors/vim.vim
+++ b/runtime/colors/vim.vim
@@ -14,8 +14,8 @@ let g:colors_name = 'vim'
 
 " General
 hi Conceal      guifg=LightGrey guibg=DarkGrey guisp=NONE gui=NONE         ctermfg=LightGrey ctermbg=DarkGrey cterm=NONE
-hi Cursor       guifg=bg        guibg=fg       guisp=NONE gui=NONE         ctermfg=bg        ctermbg=fg       cterm=NONE
-hi lCursor      guifg=bg        guibg=fg       guisp=NONE gui=NONE         ctermfg=bg        ctermbg=fg       cterm=NONE
+hi Cursor       guifg=bg        guibg=fg       guisp=NONE gui=NONE         ctermfg=NONE      ctermbg=NONE     cterm=NONE
+hi lCursor      guifg=bg        guibg=fg       guisp=NONE gui=NONE         ctermfg=NONE      ctermbg=NONE     cterm=NONE
 hi DiffText     guifg=NONE      guibg=Red      guisp=NONE gui=bold         ctermfg=NONE      ctermbg=Red      cterm=bold
 hi ErrorMsg     guifg=White     guibg=Red      guisp=NONE gui=NONE         ctermfg=White     ctermbg=DarkRed  cterm=NONE
 hi IncSearch    guifg=NONE      guibg=NONE     guisp=NONE gui=reverse      ctermfg=NONE      ctermbg=NONE     cterm=reverse
```

This will use the terminal's default cursor styling, which is probably what we want in a `'notermguicolors'` environment.

echasnovski commented:
Pushed an update to not use 0 and 15 in Normal.
I am OK to also not use them at all and use `cterm=reverse` instead, but not sure if this is a good fit for default color scheme. As this would make more breaks for Vim color scheme which don't explicitly supply `cterm=NONE`.

clason commented:
> I really think that current values have the most possible compromise between "much contrast" and "not too much contrast" camps.

Perfect! That looks (subjectively) _much_ crisper while still being not too harsh (especially in light mode). That was exactly what I was hoping for.

> not really good result inside markdown code blocks (falling back to @text.literal which links to Comment). @clason, this is probably better be addressed in tree-sitter highlight groups rework, correct?

It's a general problem that's cropped up in nvim-treesitter as well; not sure what the best way of handling it (and whether it _can_ be handled completely). Possibly it makes sense to define the tree-sitter groups independently (without links) and fully (and same for LSP semantic tokens). But yes, this would be part of looking at the tree-sitter groups and need not be considered in this PR.

famiu commented:
>    I **really** think that current values have the most possible compromise between "much contrast" and "not too much contrast" camps. All contrast ratio values passed previously (except objectively hard "saturated dark cyan") and now they pass even better (for example, `Comment` is now closer to passing with AAA than with AA).

I love the new contrast ratio much better. Well done!

>    1. `CurSearch` is changed to actually be different from `Search` (with antipodal lightness value of yellow as background). This adds another breaking behavior for color scheme authors, but seems to be worth having in default color scheme.

This is a good decision imo. Having `CurSearch` be different is actually immensely helpful in my personal usecase, at least.

>    2. `WinSeparator` was previously linked to a `VertSplit` for compatibility with Vim's color scheme. As `VertSplit` is deprecated since 0.8, I decided to revert the link. Even now I think it is a reasonable thing to do, but this might need extra discussion. Here are possible actions:
>       
>       * Leave the current `VertSplit -> WinSeparator -> Normal` linking dependency.
>       * Revert to have `WinSeparator -> VertSplit -> Normal`.
>       * Remove `VertSplit` entirely from default color scheme. Having it linked to `WinSeparator` is beneficial only to allow Neovim targeted color schemes work in Vim, which is rarely the case. So removing it completely might be reasonable here.

IMO, there is absolutely no reason to do #1. If we don't want to do #2, then we should do #3 and not #1.
Now, as for what to do between #2 and #3. The only reason to do #2, IMO, would be to have compatibility with using Vim colorscheme plugins in Neovim, which is a valid reason. It just depends on how much we care about that. I personally think removing the link is a good idea in the long term, but losing compatibility with Vim colorschemes is a price to pay for that. Then again, the user can just manually link the highlight groups, so I'm not sure if it's really as big of a deal.

echasnovski commented:
> IMO, there is absolutely no reason to do #1. If we don't want to do #2, then we should do #3 and not #1.
> Now, as for what to do between #2 and #3. The only reason to do #2, IMO, would be to have compatibility with using Vim colorscheme plugins in Neovim, which is a valid reason. It just depends on how much we care about that. I personally think removing the link is a good idea in the long term, but losing compatibility with Vim colorschemes is a price to pay for that. Then again, the user can just manually link the highlight groups, so I'm not sure if it's really as big of a deal.

Yes, I think I agree here. Will wait for more input about actually removing `VertSplit` from default color scheme.

clason commented:
> The only reason to do https://github.com/neovim/neovim/pull/2, IMO, would be to have compatibility with using Vim colorscheme plugins in Neovim, which is a valid reason.

And in fact a strong reason, as we bundle a number of very high-quality ones, which I'd hate to lose over this. Having a (commented) link from (the now canonical) `WinSeparator` to `VertSplit` is a small price to pay for that. You can move this (and possibly others) to a separate "Vim compatibility" section to make it clearer.

echasnovski commented:
> > The only reason to do #2, IMO, would be to have compatibility with using Vim colorscheme plugins in Neovim, which is a valid reason.
> 
> And in fact a strong reason, as we bundle a number of very high-quality ones, which I'd hate to lose over this. Having a (commented) link from (the now canonical) `WinSeparator` to `VertSplit` is a small price to pay for that. You can move this (and possibly others) to a separate "Vim compatibility" section to make it clearer.

They won't get lost because of that, as  there is a "Fix bundled colorschemes by adding `colorscheme vim` after `hi clear`" in the list of follow ups. Which is planned as the next step. So bundled color schemes will have `WinSeparator -> VertSplit -> Normal` links.

clason commented:
`colorscheme vim` should then set up the old default links for LSP and tree-sitter groups as well. We might actually want to think about moving these default links to a dedicated runtime script that other legacy colorschemes (or people using them) can call as well.

echasnovski commented:
Sorry, @justinmk, only now have discovered edits in [your comment about highlight group tweaks](https://github.com/neovim/neovim/issues/26369#issuecomment-1837497703).

> * (Edit): Also `Search`. I am not seeing a reason why `Search` needs to be yellow. In general "block highlights" are something that can be used to make special cases standout without introducing a new color. Example:
>  ```
>  hi Search guibg=cyan guifg=black
>  hi CurSearch guifg=black guibg=NvimDarkCyan
>  ```

Yellow is quite objectively the best possible color for `Search` and `CurSearch`. For at least these reasons:
- Yellow is reserved to "mean mild user attention", which `Search`, `CurSearch`, and `IncSearch` is one of the primary examples.
- Yellow is not used in any UI or Syntax groups (except `DiagnosticWarn` which usually can't intersect with search). So it can't possibly get in the situation of blending in the surrounding text. While this can happen if using cyan or blue for that. Here are examples:

    <details><summary>Using background lightness for `Search` background</summary>
    
    ![screenshot_2023-12-13_12:33:17](https://github.com/neovim/neovim/assets/24854248/b4d7dff0-aa21-4426-ab94-96470e75aadc)
    ![screenshot_2023-12-13_12:33:26](https://github.com/neovim/neovim/assets/24854248/f8a283d2-2344-457d-8618-f9438a1076fd)
    
    </details>

    <details><summary>Using foreground lightness for `Search` background</summary>
    
    ![screenshot_2023-12-13_12:31:56](https://github.com/neovim/neovim/assets/24854248/d95a5bb4-3948-4af5-865d-ae60b3440abd)
    ![screenshot_2023-12-13_12:32:33](https://github.com/neovim/neovim/assets/24854248/b5149970-9b7c-4670-9ae9-2698e4b1a3ee)
    
    </details>
    
    <details><summary>Behavior after this PR (yellow with background lightness for `Search` background</summary>
    
    ![screenshot_2023-12-13_12:34:12](https://github.com/neovim/neovim/assets/24854248/66433de8-22bc-4142-96ae-0a35cb5b2bed)
    ![screenshot_2023-12-13_12:34:18](https://github.com/neovim/neovim/assets/24854248/3124701e-6c9f-4354-82c7-ed72cececcae)
    
    </details>

    In both cases of using cyan, there is an instance of using very close color for background as for surrounding foreground. The only way to solve this is to not use color which can be seen in the text, i.e. green/cyan/blue. So using yellow is the best approach here.

> * (Edit): `StatusLine` should be more clearly distinct from `StatusLineNC` (probably by dimming the bg of "NC")

Unfortunately, not really possible as statusline already uses the darkest background.

I can suggest one of the two approaches here:
- Use same background as `Normal` for `StatusLineNC`, which will essentially hide statusline. Here are examples:
    <details><summary>Blended `StatusLineNC`</summary>
    
    ![screenshot_2023-12-13_12:48:34](https://github.com/neovim/neovim/assets/24854248/3256025c-0314-4833-9c59-a464cb2f3672)
    ![screenshot_2023-12-13_12:48:44](https://github.com/neovim/neovim/assets/24854248/6fbe4c96-a00d-4584-a286-3950dd1a5d54)
    
    </details>

- In case that blend is too much (which it think it is not), add `gui=bold` for both `StatusLine` and `StatusLineNC` (note that this will affect `TabLine` and `WinBar`).:
    <details><summary>Blended `StatusLineNC` and bold text in any statusline</summary>
    
    ![screenshot_2023-12-13_12:49:25](https://github.com/neovim/neovim/assets/24854248/b3bb2f28-f878-460a-91ab-509f6b69db60)
    ![screenshot_2023-12-13_12:49:31](https://github.com/neovim/neovim/assets/24854248/de975d1f-bf5d-4ac7-ac87-7361fbc89a1e)

    </details>

> * (Edit): diffAdded, diffRemoved need explicit overrides, they should be green/red
>  ```
>  hi diffAdded guifg=lightgreen
>  hi diffRemoved guifg=lightred
>  hi! link Type Comment
>  ```

I assumed this is a task for filetype syntax, i.e. ['runtime/syntax/diff.vim'](https://github.com/neovim/neovim/blob/d65c6a0bafada059e87a11a4bcd129afc16d2e5d/runtime/syntax/diff.vim#L368-L386). Is it a good idea to add them in the default color scheme?

clason commented:
Also, _please_ consider not only yourself but also other contributors, several of which (among them myself) have vision problems of some sort. Such choices can make a world of difference then. I know you have different priorities, but improving accessibility is still one of the goals of this update.

craigmac commented:
Tweaks I've made for when background is set to light:

```lua
 -- Lighten ColorColumn to fix Comment guifg being unreadable
 vim.cmd[[hi! ColorColumn guibg=NvimLightGrey1]]
 -- Lighten Visual to fix Comment guifg being unreadable
vim.cmd[[hi! Visual guibg=NvimLightGrey3]]
```

echasnovski commented:
> Tweaks I've made for when background is set to light:
> 
> ```lua
>  -- Lighten ColorColumn to fix Comment guifg being unreadable
>  vim.cmd[[hi! ColorColumn guibg=NvimLightGrey1]]
>  -- Lighten Visual to fix Comment guifg being unreadable
> vim.cmd[[hi! Visual guibg=NvimLightGrey3]]
> ```

For `Comment` on `Visual` I kind of agree that it has low contrast, but that is to be expected as with dark background `Comment` intentionally has the lowest foreground lightness and `Visual` - the highest background lightness. Having `Visual` have `Grey3` background is not really good as it is the same color as used for `CursorLine` and `CursorColumn`, which is slightly confusing. As visual selection very rarely implies actually reading text, and comments are not the most used text, and result is still somewhat readable, I consider this a reasonable compromise.

Same logic applies to `ColorColumn`, but with even less consequences as it is (usually) only single cell wide.

craigmac commented:
> For `Comment` on `Visual` I kind of agree that it has low contrast, but that is to be expected as with dark background `Comment` intentionally has the lowest foreground lightness and `Visual` - the highest background lightness. 

It's not low contrast, it's completely invisible! Here's a shot, tested on WezTerm, with neovim built from this PR:

![Screen Shot 2023-12-15 at 2 55 53 PM](https://github.com/neovim/neovim/assets/7416158/9d7f83b9-2433-45e3-ba2e-d675bf851a54)

Note this Comment/Visual issue is on for light version, the dark version is fine.

>Having `Visual` have `Grey3` background is not really good as it is the same color as used for `CursorLine` and `CursorColumn`, which is slightly confusing. 

NvimLightGrey3 guibg for Visual here at least makes the Comment guifg text visible.
I agree there could be confusion with Visual and CursorLine having same guibg, but when visual mode is started, e.g., doing `vap`, CursorLine highlights don't apply (cursorline isn't drawn). Here's the same pic as above, but now with my change for Visual mode (cursorline is on here):

![Screen Shot 2023-12-15 at 3 05 28 PM](https://github.com/neovim/neovim/assets/7416158/b64c5e6e-ada9-4c48-9484-f5029c1ff730)

> As visual selection very rarely implies actually reading text, and comments are not the most used text, and result is still somewhat readable, I consider this a reasonable compromise.

The comments when selected aren't readable at all, unless my setup is doing something odd, but I tested on Linux on macOS in a few different terminal emulators and in Neovide.
 
> Same logic applies to `ColorColumn`, but with even less consequences as it is (usually) only single cell wide.

Why is the changes in the screenshot below using `hi! ColorColumn guibg=NvimLightGrey1`  not as good as hiding the text? I think that's poor UX - no other colorscheme that is shipped with Neovim does this (look at `colorscheme shine` or `colorscheme quiet` light-version). Yes it's one character hidden, but sometimes I'll have something like `:set cc=10,20,40,60` set, and that would be a bit annoying.

![Screen Shot 2023-12-15 at 3 06 58 PM](https://github.com/neovim/neovim/assets/7416158/0ca52c71-ae8f-4bcc-90f5-eac90034a9e7)

echasnovski commented:
> It's not low contrast, it's completely invisible! Here's a shot, tested on WezTerm, with neovim built from this PR:

This seems like some issue with `WezTerm` or something else's setup. Here is the example of how this PR looks in stock `WezTerm` without config:

![screenshot_2023-12-15_22:21:52](https://github.com/neovim/neovim/assets/24854248/0e97ef9a-9451-4e86-a247-a18ceac7450d)

This can not possibly be the result of the Neovim's default color scheme. `Comment` foreground and `Visual` background use two different colors. So they should not be displayed as same color.

craigmac commented:
> This can not possibly be the result of the Neovim's default color scheme. `Comment` foreground and `Visual` background use two different colors. So they should not be displayed as same color.

Ok, ran stock Wezterm and it turned off my color tweaks and I can reproduce what you've shown there, so that's good. The odd thing I've narrowed it down to is that when I set `vim.cmd[[hi! Comment gui=italic]]` in my init.lua and I restart nvim (doesn't happen when I do it while running), `:hi Comment` gets set to `NvimLightGrey4` instead of what it should be, `guibg=NvimDarkGrey4`. Weird... how to debug further?

craigmac commented:
Ok this apparently happens with with String as well, just tried starting clean nvim then changing String to italics: `:hi! String gui=italic`, and then switch to light version doing `:set bg=light` and the green won't change, it's stays the dark variant which is way too bright:

![Screen Shot 2023-12-15 at 3 45 40 PM](https://github.com/neovim/neovim/assets/7416158/d4e28440-dcd6-4239-b6bd-eb7249e95608)

clason commented on runtime/doc/dev_theme.txt:
```diff
@@ -67,8 +66,17 @@ Palettes
+- For 16 colors:
+
+    - Greys are not used and are replaced with foreground and background of
+      terminal emulator.
```
```suggestion
      the terminal emulator's color scheme.
```

zeertzjq replied:
I think "the foreground and background **colors** of the terminal emulator" is better

clason commented on runtime/doc/dev_theme.txt:
```diff
@@ -49,8 +49,7 @@ Palettes
 
-    - Four shades of "colored" greys for general UI. In 256 colors they are
-      exact greys; in true colors they are shades of "cold" grey.
+    - Four shades of colored "cold" greys for general UI.
```
Contender for the smallest nit picked: since we use AE ("colors"), "gray" is more common than "grey" (which is mostly BE).

echasnovski replied:
Hmm... Default colors are `Nvim...Grey`. Is it an issue? Other greys are doubled with "grey" and "gray", by the way.

clason replied:
Not an issue; I just noticed the juxtaposition of "color" (AE) and "grey" (BE). As long as we are consistent, no need to support both spellings.

echasnovski replied:
As "grey" seems to be used more often than "gray" in help files (albeit pretty close), I'd stick with "grey".

clason commented on runtime/doc/news.txt:
```diff
@@ -92,9 +92,22 @@ The following changes may require adaptations in user config or plugins.
+  • |hl-WinBarNC| is linked to |hl-StatusLineNC| instead of |hl-WinBar|.
+  • |hl-WinSeparator| is linked to |hl-Normal| instead of |hl-VertSplit|.
+
+  This also might result into some color schemes looking differently.
+  It is due them relying on an implicit assumptions about how highlight groups
```
```suggestion
  This also might result into some color schemes looking differently due to them
  relying on an implicit assumptions about how highlight groups
```

clason commented on runtime/doc/news.txt:
```diff
@@ -92,9 +92,22 @@ The following changes may require adaptations in user config or plugins.
+
+  This also might result into some color schemes looking differently.
+  It is due them relying on an implicit assumptions about how highlight groups
+  are defined by default. To overcome this, define all attributes of highlight
```
```suggestion
  are defined by default. To account for this, define all attributes of highlight
```

clason commented on runtime/doc/vim_diff.txt:
```diff
@@ -30,8 +30,11 @@ Defaults					            *nvim-defaults*
   ":syntax off" to |init.vim|.
-- Default color scheme has been updated. This can be reversed by adding
-  ":colorscheme vim" to |init.vim|.
+- Default color scheme has been updated. In particular it might result into
+  color schemes looking differently in Nvim. It is due to them relying on
+  implicit assumptions of how highlight groups are defined by default.
+  This can be reversed by adding ":colorscheme vim" to |init.vim| or color
+  scheme file.
```
```suggestion
- Default color scheme has been updated. This can result in color schemes 
  looking differently due to them relying on how highlight groups are defined 
  by default. Add ":colorscheme vim" to |init.vim| or your color scheme file to
  restore the old default links and colors.
```

echasnovski commented:
> Ok, ran stock Wezterm and it turned off my color tweaks and I can reproduce what you've shown there, so that's good. The odd thing I've narrowed it down to is that when I set `vim.cmd[[hi! Comment gui=italic]]` in my init.lua and I restart nvim (doesn't happen when I do it while running), `:hi Comment` gets set to `NvimLightGrey4` instead of what it should be, `guibg=NvimDarkGrey4`. Weird... how to debug further?

I could reproduce this with set `gui` attribute. This seems to be not completely new issue: #26603.

echasnovski commented:
Addressed all docs review and updated tests to pass. All green now and, I think, ready to merge.

All other minor tweaks (specific to highlight groups, etc.) is better to be done in separate PRs.

clason commented:
Thank you for your continuing efforts on this! Merged to keep this moving forward.

Still open points (`StatusLineNC`, `Search`, `diffAdded`) are better handled as individual follow-up PRs/issues.

echasnovski commented:
Thanks for review and merge!

justinmk commented:
> I assumed this is a task for filetype syntax, i.e. ['runtime/syntax/diff.vim'](https://github.com/neovim/neovim/blob/d65c6a0bafada059e87a11a4bcd129afc16d2e5d/runtime/syntax/diff.vim#L368-L386). Is it a good idea to add them in the default color scheme?

Yes, they are very important. 

> > * (Edit): `StatusLine` should be more clearly distinct from `StatusLineNC` (probably by dimming the bg of "NC")
> 
> Unfortunately, not really possible as statusline already uses the darkest background.

Use white or something, it doesn't matter. Anything that makes `StatusLineNC` clearly different from `StatusLine`, so that the current window is obvious.

> Yellow is quite objectively the best possible color for `Search` and `CurSearch`. For at least these reasons:

I don't agree. We can use cyan block highlight. It looks fine.

echasnovski commented:
> Yes, they are very important.

I made a change to 'vim/vim' and now `diffAdded` and `diffRemoved` [are linked to `DiffAdd` and `DiffDelete`](https://github.com/neovim/neovim/blob/91d76ac941a26f8370c48e062b5e09f98c75f7bc/runtime/syntax/diff.vim#L381-L382). This translates into green background and bold red foreground respectively. Is this enough?

If this is not enough, then I suggest adding 'runtime/ftplugin/diff.lua' which applies default color schemes colors if they are not yet defined.

> Use white or something, it doesn't matter. Anything that makes StatusLineNC clearly different from StatusLine, so that the current window is obvious.

I [suggested two approaches](https://github.com/neovim/neovim/pull/26540#issuecomment-1853697069) with blended `StatusLineNC`. I personally use this in my color scheme and find that current window is obvious. Is any of the approaches good for you?

> I don't agree. We can use cyan block highlight. It looks fine.
> 
> All of the other "goals" that are getting tacked-onto this topic are secondary. The main goal is that it looks good and has the Nvim brand colors. There is no reason for `Search` to be yellow.

Well, sorry, but I don't agree with cyan background either and would like to not be associated with this change.

justinmk commented:
> vim/vim' and now `diffAdded` and `diffRemoved` [are linked to `DiffAdd` and `DiffDelete`](https://github.com/neovim/neovim/blob/91d76ac941a26f8370c48e062b5e09f98c75f7bc/runtime/syntax/diff.vim#L381-L382)

Nice!

> I [suggested two approaches](https://github.com/neovim/neovim/pull/26540#issuecomment-1853697069) with blended `StatusLineNC`. I personally use this in my color scheme and find that current window is obvious. Is any of the approaches good for you?

"Blended" is fine, but let's also use a line to separate windows. See left side for example:

![image](https://github.com/neovim/neovim/assets/1359421/efd585f0-641d-442b-bca6-2ccdec2798c8)

echasnovski commented:
> "Blended" is fine, but let's also use a line to separate windows. See left side for example:

To be clear, you suggest using `underline`, right? If yes, then it doesn't look *that* good in some terminals. Here is mine (with blended `StatusLineNC`):

![screenshot_2024-01-02_22:32:11](https://github.com/neovim/neovim/assets/24854248/47057c99-2072-4f4c-aea7-b385f44eae3f)
![screenshot_2024-01-02_22:32:27](https://github.com/neovim/neovim/assets/24854248/1c43eaa3-6f54-4484-b98f-da42362890f7)

Having it underlined to me feels like it makes harder to identify the current window.

So here are some ideas to choose from:
1. Go with blended `StatuslineNC` as suggested in my original comment.
2. Go with slightly lighter/darker background than `Normal` with a bit less contrast between `StatusLineNC` and `StatusLine`. Here are examples:

![screenshot_2024-01-02_22:30:19](https://github.com/neovim/neovim/assets/24854248/00147649-f2a6-467c-8bec-70e1a6f75e6a)
![screenshot_2024-01-02_22:30:30](https://github.com/neovim/neovim/assets/24854248/89437bb8-8988-4997-af63-567c1560f17a)

3. Go with inverted text for `StatusLine` making it *clearly* stand out. Here are examples:

![screenshot_2024-01-02_22:37:53](https://github.com/neovim/neovim/assets/24854248/f89da304-9d37-4447-8f3c-f82e5f0dd43d)
![screenshot_2024-01-02_22:38:02](https://github.com/neovim/neovim/assets/24854248/5ebbb376-a33d-4664-ae46-28b971b57621)

I personally use similar to 1, but can get behind any of these changes (preferring 2 over 3, as I don't like statusline to be the first thing that draws the eye).

-----

Side note. If `StatusLineNC` is not blended with `Normal`, then I feel there is no reason to use `underline` as colors and statusline text should already distinguish windows quite nicely.

justinmk commented:
> Having it underlined to me feels like it makes harder to identify the current window.

The current window should have fg/bg swapped, as in my screenshot.

> you suggest using `underline`, right? If yes, then it doesn't look _that_ good in some terminals

We need to accept that. Can't fix every terminal.

craigmac commented:
> To be clear, you suggest using underline, right? 

Maybe, to avoid the odd look of underlined StatusLine on some terminal emulators, we only use underline on StatusLineNC, where it's blended and we don't notice the bleeding of the colour behind the underline.

echasnovski commented:
> I made a change to 'vim/vim' and now `diffAdded` and `diffRemoved` [are linked to `DiffAdd` and `DiffDelete`](https://github.com/neovim/neovim/blob/91d76ac941a26f8370c48e062b5e09f98c75f7bc/runtime/syntax/diff.vim#L381-L382). This translates into green background and bold red foreground respectively. Is this enough?
> 
> If this is not enough, then I suggest adding 'runtime/ftplugin/diff.lua' which applies default color schemes colors if they are not yet defined.

As a follow-up, after #26981 `diffAdded`, `diffChanged`, and `DiffRemoved` groups are now linked to the new core `Added`, `Changed`, `Removed` groups. This changes comes from Vim. Colors in default color scheme are chosen to be green, cyan, and red foreground.

justinmk commented:
We should define a default `Comment` highlight for non-truecolor. Hopefully gray or dark-gray.

clason commented:
We can't, though. Non-truecolor uses terminal colors, and I don't think a gray is among the first 8.

echasnovski commented:
> We should define a default `Comment` highlight for non-truecolor. Hopefully gray or dark-gray.

That would require using either color 7 or 8. They are not used deliberately for the reasons described in the original comment:

> Both dark/light variant use only 8 colors: one background, one foreground, and 6 colors (red, green, yellow, blue, magenta, cyan). Not using shades of grey (7 for dark-ish foreground and 8 for grey) allows a UI which is more robust towards terminal color palette not designed for these greys to be used together with non-grey colors.

Thus `Comment` is left same as foreground. Can be made cyan (6) or blue (4), though.

clason commented:
If you want to try, `:hi Comment ctermfg=7` on Apple Terminal:
<img width="199" alt="Screenshot 2024-01-25 at 18 38 16" src="https://github.com/neovim/neovim/assets/2361214/81acc57d-9d22-4f16-a164-ae15f0bdaf94">

This is `ctermfg=8`:
<img width="199" alt="Screenshot 2024-01-25 at 18 38 54"
 src="https://github.com/neovim/neovim/assets/2361214/522296ce-cd27-4871-b127-f6e9f9dbc2cb">


Dark background, `ctermfg=8`:
<img width="251" alt="Screenshot 2024-01-25 at 18 40 33" src="https://github.com/neovim/neovim/assets/2361214/7854dca3-8cc3-45e3-a7d9-e0e4c32de518">

`ctermfg=7`:
<img width="251" alt="Screenshot 2024-01-25 at 18 40 49" src="https://github.com/neovim/neovim/assets/2361214/5e2ad4a2-82a9-4a3d-b379-20c9c102bd79">

I don't consider any of these acceptable, really.

justinmk commented:
> Thus `Comment` is left same as foreground

What is the harm in setting it to something from the 256-color palette? Right now it's set to nothing, that isn't useful.

> I don't consider any of these acceptable, really.

Why? The light + `ctermfg=7` and dark + `ctermfg=8` look fine to me.

echasnovski commented:
> What is the harm in setting it to something from the 256-color palette? Right now it's set to nothing, that isn't useful.

Using 256-color palette is not aligned with the original goal of having `cterm` colors be from terminal palette.

The harm is in potential to make comments not useful by having them low contrast ratio by default. We can go Vim route and make them blue.

> Why? The light + `ctermfg=7` and dark + `ctermfg=8` look fine to me.

If anything, the reverse would be more accessible:
- In examples light + `ctermfg=7` has *very* low contrast ratio.
- Many terminal color palettes define 8-15 to be the same as 0-7.

lost22git commented:
How to set transparency for default theme

justinmk commented:
> The harm is in potential to make comments not useful by having them low contrast ratio by default.

Looks fine on my terminal. If it looks fine for most people then we should do that. The goal was never to support every random terminal, the goal is to look good by default, for most users.

clason commented:
> Color theory and accessibility were never primary goals, they were secondary.

They _were_ the primary goals in my issue, just for the record.
